silkworm/node/db/access_layer.cpp

```diff
@@ -882,121 +874,261 @@ void write_last_finalized_block(RWTxn& txn, const evmc::bytes32& hash) {
+bool DataModel::read_body(const Hash& hash, BlockNum height, BlockBody& body) const {
+    return read_body(height, hash.bytes, /*read_senders=*/false, body);
+}
+
+bool DataModel::read_body(const Hash& hash, BlockBody& body) const {
+    const bool found = db::read_body(txn_, hash, body);
+    if (found) return found;
+
+    // Then search for it in the snapshots (if any)
+    const auto block_header{read_header_from_snapshot(hash)};
+    if (block_header) {
+        return read_body(block_header->number, hash.bytes, /*read_senders=*/false, body);
+    }
+
+    return false;
 }
 
-bool DataModel::read_body(const evmc::bytes32& hash, BlockBody& body) {
-    return db::read_body(txn_, hash, body);
+std::optional<evmc::bytes32> DataModel::read_canonical_hash(BlockNum height) const {
+    // Canonical chain is likely on db: first lookup there
+    auto canonical_hash{db::read_canonical_hash(txn_, height)};
+    if (canonical_hash) return canonical_hash;
+
+    // Then search for it in the snapshots (if any)
+    if (const auto block_header{read_header_from_snapshot(height)}; block_header) {
+        canonical_hash = block_header->hash();
+    }
+    return canonical_hash;
 }
 
-bool DataModel::read_canonical_block(BlockNum height, Block& block) {
-    return db::read_canonical_block(txn_, height, block);
+std::optional<BlockHeader> DataModel::read_canonical_header(BlockNum height) const {
+    // Canonical chain is likely on db: first lookup there
+    auto canonical_header{db::read_canonical_header(txn_, height)};
+    if (canonical_header) return canonical_header;
+
+    // Then search for it in the snapshots (if any)
+    return read_header_from_snapshot(height);
 }
 
-bool DataModel::has_body(BlockNum block_number, const uint8_t (&hash)[kHashLength]) {
-    return db::has_body(txn_, block_number, hash);
+bool DataModel::read_canonical_block(BlockNum height, Block& block) const {
+    const bool found = db::read_canonical_block(txn_, height, block);
+    if (found) return found;
+
+    auto block_header{read_header_from_snapshot(height)};
+    if (!block_header) return false;
+
+    const auto hash{block_header->hash()};
+    block.header = std::move(*block_header);
+
+    return read_body(height, hash.bytes, /*read_senders=*/false, block);
 }
 
-bool DataModel::has_body(BlockNum block_number, const evmc::bytes32& hash) {
-    return db::has_body(txn_, block_number, hash);
+bool DataModel::has_body(BlockNum height, HashAsArray hash) {
+    const bool found = db::has_body(txn_, height, hash);
+    if (found) return found;
+
+    return is_body_in_snapshot(height);
 }
 
-bool DataModel::has_sibling(BlockNum block_number) {
-    return db::has_sibling(txn_, block_number);
+bool DataModel::has_body(BlockNum height, const Hash& hash) {
+    return has_body(height, hash.bytes);
 }
 
 bool DataModel::read_block_by_number(BlockNum number, bool read_senders, Block& block) {
     return db::read_block_by_number(txn_, number, read_senders, block);
```

Comment:
We need to reimplement read_block_by_number here because it reads from canonical table (see comments above)